_posts/2024-04-26-modules-101.markdown

```diff
@@ -0,0 +1,135 @@
+---
+layout: post
+title:  "Valkey modules 101"
+authors: 
+  - dmitrypol
+date: 2024-04-26 01:01:01 -0700
+categories: modules
+---
+
+## What are Valkey modules?  
+
+The idea of modules is to keep the core Valkey functionality lean and add extra features (such as new commands and data types) where needed.  Modules can be written in C or other languages that have C bindings.  In this article we will go through the process of building simple modules in C and Rust (using the Valkey Module Rust SDK).  This article expects the audience to be at least somewhat familiar with C, Rust and Valkey.  
+
+## Hello World module in C
+
+If we clone [Valkey](https://github.com/valkey-io/valkey) repo we will find numerous examples in `src/modules`.  Let's create a new file `module1.c` in the same folder.  
+
+```c
+#include "../valkeymodule.h"
+
+int hello(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
+    VALKEYMODULE_NOT_USED(argv);
+    VALKEYMODULE_NOT_USED(argc);
+    return ValkeyModule_ReplyWithSimpleString(ctx, "world1");
+}
+
+int ValkeyModule_OnLoad(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
+    VALKEYMODULE_NOT_USED(argv);
+    VALKEYMODULE_NOT_USED(argc);
+    if (ValkeyModule_Init(ctx,"module1",1,VALKEYMODULE_APIVER_1) 
+        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
+    if (ValkeyModule_CreateCommand(ctx,"module1.hello", hello,"",0,0,0) 
+        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
+    return VALKEYMODULE_OK;
+}
+```
+
+All we are doing is initializing `module1` and creating a command `hello` that returns `world1`.  Now we need to update `src/modules/Makefile`
+
+```makefile
+all: ... module1.so
+
+module1.xo: ../valkeymodule.h
+
+module1.so: module1.xo
+	$(LD) -o $@ $^ $(SHOBJ_LDFLAGS) $(LIBS) -lc
+```
+
+Run `make` inside `src/modules` folder.  We will see numerous `*.so` and `*.xo` files created in `src/modules` including `module1.so`.  
```

Comment:
```suggestion
Run `make module1.so` inside `src/modules` folder.  This will compile our module in the src/modules folder. 
```

No need to generate everything.

Reply:
fair point, updated